File: backend/app/api/v1/store/orgs.py
"""Organizations REST API — membership listing (worker-owned SQLite)."""

from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.v1.store._common import get_db, get_principal
from app.db.models import Organization, OrgMembership, Profile
from app.db.scope import Principal, require_org
from app.schemas.common import BaseSchema

router = APIRouter()


# ---- schemas --------------------------------------------------------------
class OrgResponse(BaseSchema):
    id: str
    name: str
    slug: str
    role: str


class OrgMemberResponse(BaseSchema):
    user_id: str
    role: str
    email: str | None = None
    full_name: str | None = None
    avatar_url: str | None = None


# ---- routes ---------------------------------------------------------------
@router.get("/orgs", response_model=list[OrgResponse])
def list_orgs(
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> list[OrgResponse]:
    rows = session.execute(
        select(Organization, OrgMembership.role)
        .join(OrgMembership, OrgMembership.org_id == Organization.id)
        .where(OrgMembership.user_id == principal.user_id)
    ).all()
    return [
        OrgResponse(id=org.id, name=org.name, slug=org.slug, role=role)
        for org, role in rows
    ]


@router.get("/orgs/{org_id}/members", response_model=list[OrgMemberResponse])
def list_org_members(
    org_id: str,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> list[OrgMemberResponse]:
    require_org(principal, org_id)
    rows = session.execute(
        select(OrgMembership, Profile)
        .join(Profile, Profile.id == OrgMembership.user_id)
        .where(OrgMembership.org_id == org_id)
    ).all()
    return [
        OrgMemberResponse(
            user_id=m.user_id,
            role=m.role,
            email=p.email,
            full_name=p.full_name,
            avatar_url=p.avatar_url,
        )
        for m, p in rows
    ]

File: worker/internal/httpapi/orgs.go
package httpapi

import (
	"net/http"

	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/store"
	"github.com/go-chi/chi/v5"
)

// orgJSON is the wire shape (matches OrgResponse in app/api/v1/store/orgs.py).
type orgJSON struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	Slug string `json:"slug"`
	Role string `json:"role"`
}

// orgMemberJSON is the wire shape (matches OrgMemberResponse in orgs.py).
type orgMemberJSON struct {
	UserID    string  `json:"user_id"`
	Role      string  `json:"role"`
	Email     *string `json:"email"`
	FullName  *string `json:"full_name"`
	AvatarURL *string `json:"avatar_url"`
}

// RegisterOrgs mounts the orgs routes (all auth-required). Flat chi patterns;
// the org path param is {orgID} everywhere.
func (s *Server) RegisterOrgs(r chi.Router) {
	r.Get("/orgs", s.listOrgs)
	r.Get("/orgs/{orgID}/members", s.listOrgMembers)
}

func (s *Server) listOrgs(w http.ResponseWriter, r *http.Request) {
	p := principalFromCtx(r.Context())
	orgs, err := store.ListOrgs(r.Context(), s.DB, p)
	if storeError(w, err) {
		return
	}
	out := make([]orgJSON, 0, len(orgs))
	for _, o := range orgs {
		out = append(out, orgJSON{o.ID, o.Name, o.Slug, o.Role})
	}
	writeJSON(w, http.StatusOK, out)
}

func (s *Server) listOrgMembers(w http.ResponseWriter, r *http.Request) {
	p := principalFromCtx(r.Context())
	orgID := chi.URLParam(r, "orgID")
	if err := p.RequireOrg(orgID); err != nil { // 403 on foreign/unknown org
		storeError(w, err)
		return
	}
	members, err := store.ListOrgMembers(r.Context(), s.DB, orgID)
	if storeError(w, err) {
		return
	}
	out := make([]orgMemberJSON, 0, len(members))
	for _, m := range members {
		out = append(out, orgMemberJSON{m.UserID, m.Role, m.Email, m.FullName, m.AvatarURL})
	}
	writeJSON(w, http.StatusOK, out)
}
